### PR TITLE
feat: Add parameterization custom code logic

### PR DESCRIPTION
**Pull request by e-fisher**
## Description

Add logic to support parameterization rule custom code value. The snippet will be wrapped in a function `getCorrelationValueN`. This is different to custom code rule where the snippet is dropped into script without wrapping it, but I went with this approach as it more versatile and helps avoid variable name collision. Also, in the future this could be extended to pass request details to the custom code function to allow more customization, e.g.,: `getCorrelationValue0(params, headers, body)`.

The UI for adding custom code parameterization will be added in a separate PR. 

## How to Test

- Manually add parameterization rule with custom code
- Verify snippet is added to script
- Verify snippet is referenced in request
- Verify snippet function is added only once when referenced in multiple requests

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have added tests for my changes.
- [x] I have run linter locally (`npm run lint`) and all checks pass.
- [x] I have run tests locally (`npm test`) and all tests pass.
- [x] I have commented on my code, particularly in hard-to-understand areas.

## Screenshots (if appropriate):
CleanShot 2024-10-28 at 11 01 43

### CODE REVIEW

File: src/rules/shared.ts [hunk 1]
```diff
@@ -170,7 +170,7 @@ export const replaceBeginEndBody = (
     selector.begin,
     selector.end
   )
-  if (!valueToReplace) return request
+  if (!valueToReplace) return
```
Comment (e-fisher) [inline src/rules/shared.ts:173]: Previously, we had no way of knowing if the request matched when the value replaced is the same previous value. While it does not change script, it still should be displayed as matched request in rule preview. 

Returning `undefined` when there's no match allows consumers to know when rule matched.

File: src/rules/rules.ts [hunk 2]
```diff
@@ -7,10 +7,21 @@ import { createParameterizationRuleInstance } from './parameterization'
...
 import { generateSequentialInt } from './utils'
 
+function createSequentialIdPool() {
+  const currentId: Record<TestRule['type'], Generator<number>> = {
+    correlation: generateSequentialInt(),
+    parameterization: generateSequentialInt(),
+    verification: generateSequentialInt(),
+    customCode: generateSequentialInt(),
+  }
+
```
Comment (e-fisher) [inline src/rules/rules.ts:17]: Independent ID sequence for every rule type

**Review by cristianoventura — DISMISSED** [review]
I manually changed the generator to use `customCode` and the script generates as expected! 🚀

**Review by going-confetti — APPROVED** [review]
LGTM 🤩

**Pull request merged**